mammoth/model_builder.py

```diff
@@ -4,318 +4,339 @@
-from mammoth.modules.layer_stack_encoder import LayerStackEncoder
-from mammoth.modules import Embeddings
-from mammoth.modules.embeddings import PluggableEmbeddings
-from mammoth.modules.util_class import Cast
+from mammoth.inputters.vocab import Vocab
+from mammoth.models import NMTModel
+from mammoth.modules.attention_bridge import AttentionBridge
+from mammoth.modules.layer_stack import AdaptedAttentionLayersStack, StackXcoder
 from mammoth.utils.logging import logger
 from mammoth.utils.misc import use_gpu
-from mammoth.utils.module_splitter import _combine_ordered_dicts
-from mammoth.utils.parse import ArgumentParser
 
-from mammoth.modules.attention_bridge import AttentionBridge
 
+def _combine_ordered_dicts(input_dicts: Dict[str, OrderedDict]) -> OrderedDict:
+    result = []
+    for prefix, input_dict in input_dicts.items():
+        for key, item in input_dict.items():
+            result.append((f'{prefix}{key}', item))
+    return OrderedDict(result)
 
-def build_embeddings(opts, vocab, for_encoder=True):
-    """
-    Args:
-        opts: the option in current environment.
-        vocab: stoi-ish object.
-        for_encoder(bool): build Embeddings for encoder or decoder?
-    """
-    word_padding_idx = vocab.stoi[DefaultTokens.PAD]
-    opts.word_padding_idx = word_padding_idx
 
-    freeze_word_vecs = opts.freeze_word_vecs_enc if for_encoder else opts.freeze_word_vecs_dec
-    emb = Embeddings(
-        word_vec_size=opts.model_dim,
-        position_encoding=opts.position_encoding,
-        dropout=opts.dropout[0] if isinstance(opts.dropout, list) else opts.dropout,
-        word_padding_idx=word_padding_idx,
-        word_vocab_size=len(vocab),
-        freeze_word_vecs=freeze_word_vecs,
-        enable_embeddingless=opts.enable_embeddingless
-    )
-    if opts.enable_embeddingless:
-        logger.info("Creating an embeddingless model.")
-    return emb
+def uses_adapters(opts):
+    return 'adapters' in opts and opts.adapters
 
 
-def build_encoder(opts, embeddings, task_queue_manager):
-    """
-    Various encoder dispatcher function.
-    Args:
-        opts: the option in current environment.
-        embeddings (Embeddings): vocab embeddings for this encoder.
+def get_attention_layers_kwargs(
+    side: Side,
+    layer_stack_index,
+    xcoder_id,
+    model_opts,
+):
+    """Return arguments for x_transformers.AttentionLayers"""
+    assert side in {Side.encoder, Side.decoder}, f'Invalid side "{side}"'
+    depths = model_opts.enc_layers if side == Side.encoder else model_opts.dec_layers
+    depth = depths[layer_stack_index]
+    causal = side == Side.decoder
+    cross_attend = side == Side.decoder
+    is_last = layer_stack_index == len(depths) - 1
+    # changed from default
+    use_simple_rmsnorm = True
+    attn_flash = True
+    ff_glu = True
+    pre_norm_has_final_norm = is_last
+    # Mostly x_transformers defaults. Make (some of this) configurable.
```

Comment:
couldn't we pass from the opts and unpack here a global `--x_transformer_config` ?

Reply:
That might be nice.

Pros:
- Less code to write and maintain.
- Very clear from the config which opts correspond to the ones in x-transformers.

Cons:
- Less discoverable, because the pass-through opts are not mentioned in our `opts.py`.
- Some opts need to be used also outside of x-transformers, and some opts don't work and need to be prevented through validation.

Reply:
> - Less code to write and maintain.

This is necessary, if we want the lib to be maintainable by less than 1 person... 

> - Less discoverable, because the pass-through opts are not mentioned in our opts.py.

I agree, but the whole point of this xtransformers business is to circumscribe what's our maintenance responsibility. a good point to the xtransformers docs should be enough; if you feel overzealous, consider mentioning the few key args in our doc as well

> - Some opts need to be used also outside of x-transformers, and some opts don't work and need to be prevented through validation.

True, but it wouldn't be hard to add an assert at the end of parsing / args validation for the few keys you need from x-transformers

Reply:
Implemented

Reply:
> a good point to the xtransformers docs should be enough

Unfortunately, x-transformers does not document these well. I linked to two places: the code where all options are listed (name only), and the readme which presents some highlights with explanation and citation in long form.